### PR TITLE
Pass schema in credentials to Postgresql

### PR DESCRIPTION
**Pull request by buremba**
There is a property called `schema` in Postgresql credentials but it looks like DBT doesn't actually use it probably because Psycopg2 doesn't provide a native way to pass `schema` to Postgresql. 

It's a bit *tricky* but there's actually a way to do that via `options` of libpq. I believe that we should support this feature since we have the `schema` parameter in Postgresql configuration.

### CODE REVIEW

**Comment by drewbanin** [comment]
Hi @buremba - thanks for making this PR! Can you tell me more about why you're interested in setting the search path for the Postgres connection? This isn't something I've ever had to do before, and I'm curious what motivated this change on your end.

**Comment by buremba** [comment]
Hey @drewbanin, thanks for the answer. We use DBT within our product, the users often write SQL queries without referencing the actual schema since we have the concept of default schema and most of them actually use single schema. Requiring the schema in all table references is verbose in our case.

**Comment by drewbanin** [comment]
hey @buremba - thanks for the additional information. Cool to hear that you're using dbt in your product! While the change you're describing sounds reasonable, I don't think dbt's connection logic is necessarily an appropriate place for this configuration.

Is it instead possible to use a pre-hook to run:
```
set search_path to '<search path>'
```

or can you maybe configure user accounts with:
```
ALTER ROLE <user> SET search_path = '<search path>'
```

**Comment by buremba** [comment]
@drewbanin both solutions that you suggested work in practice but I believe that they're kinda workaround considering dbt already provides a way to set the schema via credentials.

Is there any other place that you use the `schema` in this case? If not, I think that we need to remove that parameter in order to avoid confusion if we're not going to merge this PR. However; since the other adapters such as Snowflake already support this feature using the method in this PR, it would be much better to support it in PG & Redshift for convenience.

**Comment by beckjake** [comment]
@buremba We use the `schema` in dbt all over the place - if you don't specify an override `schema` to your dbt models, we use the one assigned in the credentials.

**Comment by buremba** [comment]
@beckjake, could you please point where the `schema` of `PostgrestCredential` being used in the codebase?

Let me explain the use-case again:
We use both Snowflake and Postgresql and use `schema` parameter in both adapters. When we execute a simple query such as `SELECT count(*) my_table` in Snowflake, dbt automatically execute the query in the schema that I defined since it passes schema to Snowflake.
However; that's not the case in Postgresql, `SELECT count(*) my_table` references `public.my_table` no matter what I use as `schema` in my credentials.

**Comment by drewbanin** [comment]
@buremba the `schema` config in a `profiles.yml` target fundamentally configures the destination schema that dbt builds models into. This is the case on every database that dbt supports. I would advise against removing it from the target config, as that would cause dbt to no longer work.

I'm open to discussing this feature request, and the appropriate place to do that is in an issue. Please open a [new issue](https://github.com/fishtown-analytics/dbt/issues/new) indicating that you have a feature request to supply a value for the `search_path` in a Postgres/Redshift connection. This issue template does a good job of ensuring that feature requests are well defined. I will be more than happy to discuss the implementation details with you in that issue.

> beckjake, could you please point where the schema of PostgrestCredential being used in the codebase?

Don't do that, man. Jake's written tens of thousands of lines of code in the dbt codebase - that doesn't mean that he's always right, but it does mean that you should give some thought to his feedback in a PR comment.

It sounds to me like you are using dbt in a specific way inside of your product. I am very happy to hear that! I also need you to understand that this is not the most common use-case for dbt, and the code that you've supplied in this PR would absolutely break people's dbt projects on Postgres/Redshift. I am happy to explain to you why this is the case, and I am also happy to suggest alternative approaches that will help you accomplish your goals while still ensuring that we don't break dbt projects for no reason.

I look forward to following up with you on your feature request. Closing.

**Comment by buremba** [comment]
@drewbanin @beckjake, sorry for the misunderstanding. I really thought that I was missing something so that's why I explained the case after that question.

**Comment by drewbanin** [comment]
Discussed in https://github.com/fishtown-analytics/dbt/issues/1477

**Comment by buremba** [comment]
Hey @drewbanin, it looks like the tests are failing even though I was able to make it work in my local environment. Somehow, it doesn't recognize the `search_path` property of `PostgresCredentials` even though it's already there, maybe it's related to a dependency issue? Do you mind taking a look at it, I'm kinda confused?

**Comment by drewbanin** [comment]
Hmmm, yeah, this test failure is pretty perplexing. I'll check it out!

**Comment by beckjake** [comment]
If the value isn't `required`, which it's not (as it shouldn't be!) you can't use `.search_path` to look it up. You'll want to do something like:
```
search_path = credentials.get('search_path')
if search_path:
   ...
```

Instead of the current:
```
if credentials.search_path is not None and credentials.search_path != ''
```

Which will only work if the target has a `search_path` set.

**Comment by drewbanin** [comment]
Just kicked off a build of the integration tests

File: .gitignore [hunk 1]
```diff
@@ -76,3 +76,6 @@ target/
 
 # Vim
 *.sw*
+
```
Comment (drewbanin) [inline .gitignore]: can you remove this? you can create a `.gitignore` locally without including it in this repo's `.gitignore`. Instructions here: https://help.github.com/en/articles/ignoring-files#create-a-global-gitignore

File: plugins/postgres/dbt/adapters/postgres/connections.py [hunk 2]
```diff
@@ -105,6 +108,12 @@ def open(cls, connection):
         if keepalives_idle:
             kwargs['keepalives_idle'] = keepalives_idle
 
+        # psycopg2 doesn't support search_path officially, see https://github.com/psycopg/psycopg2/issues/465
```
Comment (drewbanin) [inline plugins/postgres/dbt/adapters/postgres/connections.py]:
```
plugins/postgres/dbt/adapters/postgres/connections.py:111:80: E501 line too long (109 > 79 characters)
```

File: plugins/postgres/dbt/adapters/postgres/connections.py [hunk 3]
```diff
@@ -105,6 +108,12 @@ def open(cls, connection):
...
+        search_path = credentials.get('search_path')
+        if search_path is not None and search_path != '':
+            # see https://www.postgresql.org/docs/9.5/libpq-connect.html#LIBPQ-CONNECT-OPTIONS
```
Comment (drewbanin) [inline plugins/postgres/dbt/adapters/postgres/connections.py]:
```
plugins/postgres/dbt/adapters/postgres/connections.py:114:80: E501 line too long (94 > 79 characters)
```

File: plugins/postgres/dbt/adapters/postgres/connections.py [hunk 4]
```diff
@@ -105,6 +108,12 @@ def open(cls, connection):
...
+            kwargs['options'] = '-c search_path={}'.format(search_path.replace(' ', '\\ '))
```
Comment (drewbanin) [inline plugins/postgres/dbt/adapters/postgres/connections.py]:
```
plugins/postgres/dbt/adapters/postgres/connections.py:115:80: E501 line too long (91 > 79 characters)
```

File: plugins/redshift/dbt/adapters/redshift/connections.py [hunk 5]
```diff
@@ -75,7 +78,7 @@ def type(self):
         return 'redshift'
 
     def _connection_keys(self):
-        return ('host', 'port', 'user', 'database', 'schema', 'method')
+        return ('host', 'port', 'user', 'database', 'schema', 'method', 'search_path')
```
Comment (drewbanin) [inline plugins/redshift/dbt/adapters/redshift/connections.py]:
```
plugins/redshift/dbt/adapters/redshift/connections.py:81:80: E501 line too long (86 > 79 characters)
```

**Comment by beckjake** [comment]
This looks great to me once the lint issues are resolved

**Comment by buremba** [comment]
It looks like the test failures are not related to the changes done in this PR. Can someone review the PR again, please?

**Comment by drewbanin** [comment]
Just kicked off the tests again - looks like there's one remaining flake8 issue:
```
plugins/redshift/dbt/adapters/redshift/connections.py:130:21: E131 continuation line unaligned for hanging indent
```

**Comment by buremba** [comment]
Looks like PyCharm is not fully compatible with PEP8 as `Format Code` caused another PEP8 issue. 😬
Is there any way for me to kick off the tests for a PR so that I don't take your time?

**Comment by beckjake** [comment]
You should be able to just run `flake8 --select=E,W,F --ignore=W504 core/dbt plugins/*/dbt` after installing dbt's `dev_requirements.txt`

**Comment by buremba** [comment]
@beckjake yep, that's what I did after seeing it in`tox.ini`. However; your CI environment is doing much more than that (`.circleci/config.yml`) so it would be great it's triggered automatically.

**Comment by beckjake** [comment]
CircleCI is just running flake8 via tox, as specified [here](https://github.com/fishtown-analytics/dbt/blob/73607b85b7825b7fca01a340c52469317eaeae7a/.circleci/config.yml#L22) - so that flake8 command should be it.

**Comment by buremba** [comment]
I assumed that the jobs for integration tests in that file are also active, aren't they? I will be running `flake8` before pushing commits to the PRs but if I can be sure that all the tests are passing just fine before requesting a review, it would be much more efficient for both of us IMO.

**Comment by beckjake** [comment]
Oh, yes, that will only test the flake8 part. The azure pipeline tests will run for unit and postgres integration tests without us interacting, I think if those + local flake8 tests pass the easiest thing is to wait for us to run them - running the integration tests locally requires a snowflake/redshift/bigquery account, which is why we have to kick those off manually: we don't want to expose authentication information to pull requests without verifying the code.

For this change, I think if your postgres + unit + flake8 tests all pass, you are probably going to be ok!

**Comment by buremba** [comment]
FYI, my last commit fixes the PEP8 Code Style issue so this PR should be good to go.

**Comment by drewbanin** [comment]
Thanks for the update @buremba - this is going to get merged for 0.14.0!

We're working through some issues with our CI setup. Fear not though, this is definitely making it into our next release :)

**Comment by beckjake** [comment]
This looks good to me, thanks @buremba!

**Pull request merged**